core/store/src/db.rs

```diff
@@ -0,0 +1,187 @@
+{
+    io::Error::new(io::ErrorKind::Other, e)
+}
+
+pub struct DBTransaction {
+    pub ops: Vec<DBOp>,
+}
+
+pub enum DBOp {
+    Insert { col: usize, key: Vec<u8>, value: Vec<u8> },
+    Delete { col: usize, key: Vec<u8> },
+}
+
+impl DBTransaction {
+    pub fn put<K: AsRef<[u8]>, V: AsRef<[u8]>>(&mut self, col: usize, key: K, value: V) {
+        self.ops.push(DBOp::Insert {
+            col,
+            key: key.as_ref().to_owned(),
+            value: value.as_ref().to_owned(),
+        });
+    }
+
+    pub fn delete<K: AsRef<[u8]>>(&mut self, col: usize, key: K) {
+        self.ops.push(DBOp::Delete { col, key: key.as_ref().to_owned() });
+    }
+}
+
+pub struct RocksDB {
+    db: RwLock<Option<DB>>,
+    cfs: Vec<*const ColumnFamily>,
+    read_options: ReadOptions,
+}
+
+unsafe impl Send for RocksDB {}
+unsafe impl Sync for RocksDB {}
+
+pub struct TestDB {
+    db: RwLock<Option<Vec<HashMap<Vec<u8>, Vec<u8>>>>>,
+}
+
+pub trait Database: Sync + Send {
+    fn transaction(&self) -> DBTransaction {
+        DBTransaction { ops: Vec::new() }
+    }
+    fn get(&self, col: usize, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error>;
+    fn iter<'a>(&'a self, column: usize) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a>;
+    fn write(&self, batch: DBTransaction) -> io::Result<()>;
+}
+
+impl RocksDB {
+    fn iterate(&self, col: usize) -> Option<DBIterator<'_>> {
+        match *self.db.read() {
+            Some(ref db) => unsafe {
+                let cf_handle = &*self.cfs[col];
+                let iterator =
+                    db.iterator_cf_opt(cf_handle, &self.read_options, IteratorMode::Start).unwrap();
+                Some(iterator)
+            },
+            _ => None,
+        }
+    }
+}
+
+impl Database for RocksDB {
+    fn get(&self, col: usize, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error> {
+        match *self.db.read() {
+            Some(ref db) => unsafe {
+                Ok(db.get_cf_opt(&*self.cfs[col], key, &self.read_options).map_err(other_io_err)?)
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn iter<'a>(&'a self, col: usize) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+        Box::new(RocksDB::iterate(self, col).into_iter().flat_map(|i| i))
+    }
+
+    fn write(&self, transaction: DBTransaction) -> io::Result<()> {
+        let mut batch = WriteBatch::default();
+        for op in transaction.ops {
+            match op {
+                DBOp::Insert { col, key, value } => unsafe {
+                    batch.put_cf(&*self.cfs[col], key, value).map_err(other_io_err)?;
+                },
+                DBOp::Delete { col, key } => unsafe {
+                    batch.delete_cf(&*self.cfs[col], key).map_err(other_io_err)?;
+                },
+            }
+        }
+        match *self.db.write() {
+            Some(ref db) => Ok(db.write(batch).map_err(other_io_err)?),
+            _ => panic!(),
```

Comment:
Should probably have more explicit error message when we fail to take the lock

Reply:
it's unclear to me when would lock fails with the parking_lot's RwLock: https://docs.rs/lock_api/0.3.1/lock_api/struct.RwLock.html#method.read 

If it's the std RwLock, lock only fails when some thread hold the writer lock and panic, which is not possible, as only place of someone hold a write lock is here:
![image](https://user-images.githubusercontent.com/13259400/69579123-bf6f5400-0f86-11ea-9790-7e97b8ff9848.png)
and db.write rust-rocksdb doesn't panic, on doc and as i checked the source, `try_ffi!` doesn't panic: https://docs.rs/rocksdb/0.13.0/src/rocksdb/db.rs.html#927